visionmetrics/key_value_pair/key_value_pair_eval.py

```diff
@@ -0,0 +1,314 @@
+from enum import Enum
+import logging
+
+from visionmetrics.key_value_pair.key_value_pair_eval_base import KeyValuePairEvaluatorBase, SupportedKeyWiseMetric
+
+logger = logging.getLogger(__name__)
+
+
+class JSONSchemaKeyType(str, Enum):
+    String = "string"
+    Number = "number"
+    Integer = "integer"
+    Boolean = "boolean"
+    Array = "array"
+    Object = "object"
+
+
+# Constants for schemas
+SIMPLE_KEY_TYPES = [JSONSchemaKeyType.String, JSONSchemaKeyType.Number, JSONSchemaKeyType.Integer, JSONSchemaKeyType.Boolean]
+OUT_OF_DISTRIBUTION_ENUM_KEY = "<|other|>"
+
+# Constants for predictions and targets
+VALUE_SUBKEY = "value"
+GROUNDINGS_SUBKEY = "groundings"
+DESCRIPTION_SUBKEY = "description"
+
+
+class KeyValuePairExtractionScore(KeyValuePairEvaluatorBase):
+    """
+    Evaluator for a key-value pair dataset. The default evaluator (if the key does not have values of a particular form) is the caption.AzureOpenAITextModelCategoricalScore due to its flexibility.
+    Args:
+        key_value_pair_schema: dictionary in JSON Schema format indicating each key and expected value type for each extracted field.
+        Examples (corresponding to the examples in key_value_pair_eval_base.py):
+        defect_detection_schema = {
+            "defects": {
+                "type": "array",
+                "description": "The defect types present in the image.",
+                "items": {
+                    "type": "string",
+                    "description": "The type of defect detected",
```

Comment:
I get the pt that you want to pass the existing schema from MMI api or vision-datasets directly here, but a schema for this evaluation class should be designed from this class's perspective, not restricted by or for the convenience of the downstream application.

Example: if MMI api changes the schema format, do you want to change this library as well with the justification being MMI api updates it?

Reply:
Good question. The rationale for using the same schema is not *because* it matches the downstream application's, but moreso that the schema needs to be flexible enough to represent the task of key-value pair extraction with additional returned values like grounding for each key; so, the design is the same because the application's design also took these same factors into account to properly represent the task. Previous versions of the schema would not have enabled the correct logic for inferring metric types. JSON Schema is also a pretty standard format across various applications today.

It is true that the library will likely evolve, but the hope is that the schema is flexible enough to adapt to additional features that we would want to add for this metric.

Reply:
"JSON Schema is also a pretty standard format across various applications today."

Not really talking about whether it should be JSON or XML for the schema, but more about what the schema is, for what purpose. For example, is "description" in your schema needed for evaluation? Not saying you should fail the evaluation if "description" exists, but it shouldn't be mentioned/expected from an evaluation class.

Basically what I am asking is to check if everything from the schema in your example makes sense to evaluation. This is an evaluation class at the end of day. 

If the shared schema includes tons of dataset details not relevant to evaluation, a separation would save you from confusing library users and maintenance burden.

Reply:
JSON Schema refers not to the fact that it is in JSON format, but specifically this schema definition, which is a standard for API swaggers and also structured data extraction nowadays: https://json-schema.org/understanding-json-schema/reference.

Good point - the "description" key, for instance, is not actually used in evaluation, only provided in the examples. I'll clean up the examples so they only contain the minimal information needed for the evaluation.

Reply:
Hi @pj-ms, please take a look at the most recent code, which addresses this issue. Thanks!